### PR TITLE
[GSoC18] Refactor DataContainer and Fixed Copying Issues

### PR DESCRIPTION
**Pull request by thmq**

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/content/Project.java [hunk 1]
```diff
@@ -133,41 +131,14 @@ public Project(SupportProject supportProject, Context context) {
...
-			Iterator variablesIterator = dataContainer.spriteVariables.keySet().iterator();
-			while (variablesIterator.hasNext()) {
-				Sprite sprite = (Sprite) variablesIterator.next();
-				if (sprite == null) {
-					variablesIterator.remove();
-				}
-			}
-		}
-	}
-
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/content/Project.java]: There are ~70 programs having `null` as key. This code has been added in a hotfix https://github.com/Catrobat/Catroid/commit/626e5fdb4ca60bf6f7835301defcbab475a64c53, maybe that's why the number is so low.

Should we keep it? Maybe use [Map.remove](https://developer.android.com/reference/java/util/Map.html#remove(java.lang.Object))?
This code is a good indicator that the DataContainer should prevent from storing null Sprites (future improvements :wink:)

Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/content/Project.java]: sure, I can do that. btw: https://github.com/Catrobat/Catroid/pull/2864/files#diff-cb1b43b9b2d00fb9b796dba270e0093e prevents you from adding UserData to a null key.

Comment (robertpainsi) [reply catroid/src/main/java/org/catrobat/catroid/content/Project.java]: Awesome! :party:

File: catroid/src/main/java/org/catrobat/catroid/stage/ShowTextActor.java [hunk 2]
```diff
@@ -92,14 +91,9 @@ public static String convertToEnglishDigits(String value) {
...
-		Map<Sprite, List<UserVariable>> spriteVariableMap = dataContainer.getSpriteVariableMap();
-		List<UserVariable> spriteVariableList = spriteVariableMap.get(sprite);
-		List<UserVariable> userBrickVariableList = dataContainer.getOrCreateVariableListForUserBrick(userBrick);
-
-		drawVariables(projectVariableList, batch);
-		drawVariables(spriteVariableList, batch);
-		drawVariables(userBrickVariableList, batch);
+		drawVariables(dataContainer.getProjectUserVariables(), batch);
+		drawVariables(dataContainer.getSpriteUserVariables(sprite), batch);
+		drawVariables(dataContainer.getUserBrickUserVariables(userBrick), batch);
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/stage/ShowTextActor.java]: `java.lang.NullPointerException: This map does not allow null keys`

File: catroid/src/main/java/org/catrobat/catroid/content/Project.java (hunk 1 above)
Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/content/Project.java]: I have moved the null entry removal just below that [that] (I wasn't able to get the exact line - I'm sure I could have but it took me too long).

File: catroid/src/main/java/org/catrobat/catroid/stage/ShowTextActor.java (hunk 2 above)
Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/stage/ShowTextActor.java]: for now I went with option 2.

**Comment by thmq** [comment]
please run espresso tests

**Comment by thmq** [comment]
please run espresso tests

**Comment by thmq** [comment]
please run espresso tests

**Comment by robertpainsi** [comment]
@84n4n4, thanks for also having a look at this!

**Pull request merged**